### PR TITLE
config_settings and config_options

### PR DESCRIPTION
**Pull request by memsharded**
This removes the counter_config hack for header_only options

### CODE REVIEW

File: conans/model/conan_file.py [hunk 1]
```diff
@@ -156,7 +156,20 @@ def system_requirements(self):
...
+        """ modify options, probably conditioned to some settings. This call is executed
+        before config_settings. E.g.
+        if self.settings.os == "Windows":
+            del self.options.shared  # shared/static not supported in win
+        """
+    def config_settings(self):
+        """ modify settings, probably conditioned to some options. This call is executed
+        after config_options. E.g.
+        if self.options.header_only:
+            self.settings.clear()
```
Comment (lasote) [inline conans/model/conan_file.py]: Maybe recommend to add/remove requires in config_settings method

**Comment by lasote** [comment]
Good fix. Some test missing.

**Comment by lasote** [comment]
Why the config rename to configure? I like the change of remove config_settings, but I don't get the reason for not to keep the config name.

**Comment by memsharded** [comment]
Because it not only configures settings, but it is also the place for requirements? You pointed that out, but now I realize it is not correct, as we have the requirements() method for that. Yes, I can recover the previous config_settings name if you want.

**Comment by lasote** [comment]
I was wondering if we can add just one new method "config_options" and keep the old "config". It has sense? Could it work?

**Pull request merged**